CSharpSourceCode/CampaignSupport/TORSkillBookCampaignBehavior.cs

```diff
@@ -0,0 +1,244 @@
+                return;
+            }
+
+            var startProgression = _readingProgress.GetValueOrDefault(CurrentBook, 0);
+            var endProgression = Math.Min(startProgression + hours, hoursToComplete);
+            // If we ever decide to progress reading by > 1 hour, we should cap
+            // progression or too many skill points might be allocated.
+            hours = endProgression - startProgression;
+
+            _currentSkillTuples.ForEach(
+                skillTuple => CalculateSkillTupleProgression(skillTuple, startProgression, hours));
+
+            _readingProgress[CurrentBook] =
+                Math.Min(endProgression, hoursToComplete);
+            TOWCommon.Say(
+                String.Format("Reading for {0} has progressed to {1:0.00}%",
+                _currentBookObject.Name, 100 * ((float) endProgression / hoursToComplete)));
+        }
+
+        private void CalculateSkillTupleProgression(SkillTuple skillTuple, int currentProgression, int hours)
+        {
+            if (currentProgression >= skillTuple.LearningTime)
+                return;
+
+            // If this is an ability scroll, we don't need to do complex calculations
+            if (skillTuple.IsAbility && IsUsefulForAbility(skillTuple))
+            {
+                var abilityTemplate = AbilityFactory.GetTemplate(skillTuple.SkillId);
+                if (currentProgression + hours >= skillTuple.LearningTime
+                    && abilityTemplate != null
+                    && !Hero.MainHero.HasAbility(skillTuple.SkillId))
+                {
+                    Hero.MainHero.AddAbility(skillTuple.SkillId);
+                    TOWCommon.Say(String.Format("{0} has gained the {1} ability!", 
+                        Hero.MainHero.Name, abilityTemplate.Name));
+                }
+                return;
+            }
+
+            if (!IsUsefulForSkill(skillTuple))
+                return;
+
+            // Figure out how many skill points should be allocated for the # hours
+            // passed.
+            float currentRatio = currentProgression / skillTuple.LearningTime;
+            float progressedRatio = Math.Min(currentProgression + hours, skillTuple.LearningTime) / skillTuple.LearningTime;
+            int startGrantedSkillPoints = (int) (skillTuple.FlatSkillModifier * currentRatio);
+            int endGrantedSkillPoints = (int) (skillTuple.FlatSkillModifier * progressedRatio);
+            int skillsGained = endGrantedSkillPoints - startGrantedSkillPoints;
+
+            SkillObject skillObject = GetSkillObject(skillTuple.SkillId);
+            // TODO(jason): Replace set skill logic with add skill exp so that players
+            // also benefit from general exp.
+            Hero.MainHero.HeroDeveloper.SetInitialSkillLevel(skillObject, Hero.MainHero.GetSkillValue(skillObject) + skillsGained);
+            Hero.MainHero.HeroDeveloper.InitializeSkillXp(skillObject);
```

Comment:
yes, as it is said in the TODO, you can simply use Hero.AddSkillXp (NOT HeroDeveloper, also has a method with the same name) -> this will add normal levelup xp as well and also take care of any event notifications etc.
You also don't need a learinglimit then, because the base game's limit will be taken into account.

Reply:
makes things easier I agree.

Reply:
Oo I didn't think about the other reasons for doing so, but yea it should simplify the logic and work out better :0

I'll update the PR after work today